src/io/io.cpp

```diff
@@ -1563,37 +1563,51 @@ void Grid3D::Write_Projection_HDF5(hid_t file_id)
     dataspace_xz_id = H5Screate_simple(2, dims, NULL);
 
     // Copy the xy density and temperature projections to the memory buffer
-    for (j = 0; j < H.ny_real; j++) {
-      for (i = 0; i < H.nx_real; i++) {
+    for (int j = 0; j < H.ny_real; j++) {
+      for (int i = 0; i < H.nx_real; i++) {
         dxy = 0;
         Txy = 0;
   #ifdef DUST
         dust_xy = 0;
   #endif
         // for each xy element, sum over the z column
-        for (k = 0; k < H.nz_real; k++) {
-          id = (i + H.n_ghost) + (j + H.n_ghost) * H.nx + (k + H.n_ghost) * H.nx * H.ny;
+        for (int k = 0; k < H.nz_real; k++) {
+          int const xid = i + H.n_ghost;
+          int const yid = j + H.n_ghost;
+          int const zid = k + H.n_ghost;
+          int const id  = cuda_utilities::compute1DIndex(xid, yid, zid, H.nx, H.ny);
+
           // sum density
           dxy += C.density[id] * H.dz;
   #ifdef DUST
           dust_xy += C.dust_density[id] * H.dz;
   #endif
           // calculate number density
-          n = C.density[id] * DENSITY_UNIT / (mu * MP);
+          Real const n = C.density[id] * DENSITY_UNIT / (mu * MP);
+
   // calculate temperature
-  #ifndef DE
-          Real mx = C.momentum_x[id];
-          Real my = C.momentum_y[id];
-          Real mz = C.momentum_z[id];
-          Real E  = C.Energy[id];
-          T       = (E - 0.5 * (mx * mx + my * my + mz * mz) / C.density[id]) * (gama - 1.0) * PRESSURE_UNIT / (n * KB);
-  #endif
   #ifdef DE
-          T = C.GasEnergy[id] * PRESSURE_UNIT * (gama - 1.0) / (n * KB);
-  #endif
-          Txy += T * C.density[id] * H.dz;
+          Real const T = hydro_utilities::Calc_Temp_DE(C.density[id], C.GasEnergy[id], gama, n);
+  #else  // DE is not defined
+          Real const mx = C.momentum_x[id];
+          Real const my = C.momentum_y[id];
+          Real const mz = C.momentum_z[id];
+          Real const E  = C.Energy[id];
+
+    #ifdef MHD
+          auto const [magnetic_x, magnetic_y, magnetic_z] =
+              mhd::utils::cellCenteredMagneticFields(C.host, id, xid, yid, zid, H.n_cells, H.nx, H.ny);
+    #else   // MHD is not defined
+          Real const magnetic_x = 0.0, magnetic_y = 0.0, magnetic_z = 0.0;
+    #endif  // MHD
+
+          Real const T = hydro_utilities::Calc_Temp_Conserved(E, C.density[id], mx, my, mz, gama, n, magnetic_x,
+                                                              magnetic_y, magnetic_z);
+  #endif    // DE
+
+          Txy += T * H.dz;
```

Comment:
This line is missing the density-weighting (Cholla outputs density-weighted temperature projections)

Reply:
In that case I have a question. `hydro_utilities::Calc_Temp_DE` says that the DE temperature is 

`Real T = d * ge * (gamma - 1.0) * PRESSURE_UNIT / (n * KB);`

and the projection code has 

`Real T = C.GasEnergy[id] * PRESSURE_UNIT * (gama - 1.0) / (n * KB);`

Which differ by a factor of `d`. Is that a bug in `Calc_Temp_DE` or the projections?

Reply:
ge is usually defined as GasEnergy / density

Reply:
Ok, my read is that the `Calc_Temp_DE` version is correct and the projection version should have the additional factor of `d` in it?

Reply:
Both are the same and are correct (both versions give you T). So yes, the projection version should be multiplied by density when you do the sum (not in the line where you calculate T). If you look at the previous version of the file, it's there, I just noticed you deleted it.

Reply:
The quantity GasEnergy in the conserved array is d*ge. Does that make sense?

Reply:
Per an offline discussion with @evaneschneider, the confusion is in nomenclature. `GasEnergy` is the total thermal energy in a cell, `ge` is the specific thermal energy in a cell and they relate via the equation `GasEnergy = ge*density`.